pkg/collector/corechecks/gpu/nvidia/gpm.go

```diff
@@ -0,0 +1,237 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux && nvml
+
+package nvidia
+
+import (
+	"fmt"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+
+	"github.com/hashicorp/go-multierror"
+
+	ddnvml "github.com/DataDog/datadog-agent/pkg/gpu/safenvml"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+)
+
+const sampleBufferSize = 2
+
+type gpmCollector struct {
+	lib                 ddnvml.SafeNVML
+	device              ddnvml.SafeDevice
+	samples             []nvml.GpmSample
+	metricsToCollect    map[nvml.GpmMetricId]gpmMetric
+	nextSampleToCollect int
+}
+
+type gpmMetric struct {
+	name       string
+	metricType metrics.MetricType
+}
+
+var allGpmMetrics = map[nvml.GpmMetricId]gpmMetric{
+	nvml.GPM_METRIC_GRAPHICS_UTIL: {
+		name:       "gr_engine_active",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_SM_UTIL: {
+		name:       "sm_active",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_SM_OCCUPANCY: {
+		name:       "sm_occupancy",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_INTEGER_UTIL: {
+		name:       "integer_active",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_FP16_UTIL: {
+		name:       "fp16_active",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_FP32_UTIL: {
+		name:       "fp32_active",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_FP64_UTIL: {
+		name:       "fp64_active",
+		metricType: metrics.GaugeType,
+	},
+	nvml.GPM_METRIC_ANY_TENSOR_UTIL: {
+		name:       "tensor_active",
+		metricType: metrics.GaugeType,
+	},
+}
+
+func newGPMCollector(device ddnvml.SafeDevice) (c Collector, err error) {
+	support, err := device.GpmQueryDeviceSupport()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get GPM support: %w", err)
+	}
+
+	if support.IsSupportedDevice == 0 {
+		return nil, errUnsupportedDevice
+	}
+
+	// Clone the global allGpmMetrics map to avoid mutating global state
+	clonedMetrics := make(map[nvml.GpmMetricId]gpmMetric, len(allGpmMetrics))
+	for key, value := range allGpmMetrics {
+		clonedMetrics[key] = value
+	}
+
+	collector := &gpmCollector{
+		device:           device,
+		metricsToCollect: clonedMetrics,
+	}
+
+	collector.lib, err = ddnvml.GetSafeNvmlLib()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get NVML library: %w", err)
+	}
+
+	defer func() {
+		if err != nil {
+			// return all allocated samples to NVML if we fail after they have been allocated
+			collector.freeSamples()
+		}
+	}()
+
+	for i := 0; i < sampleBufferSize; i++ {
+		sample, err := collector.lib.GpmSampleAlloc()
+		if err != nil {
+			return nil, fmt.Errorf("failed to allocate GPM sample: %w", err)
+		}
+		collector.samples = append(collector.samples, sample)
+	}
+
+	if err := collector.removeUnsupportedMetrics(); err != nil {
+		return nil, fmt.Errorf("failed to remove unsupported metrics: %w", err)
```

Comment:
do we need to bail out here?

Reply:
i prefer a more consistent behavior between our collectors, i think rest of them don't return any error in the removeUnsupportedMetrics func (apart from fields collector).

let's make a decision here and align all collectors. I personally think that it should not return any error and be a best-effort for optimization purposes. wdyt?